execution-engine/common/src/contract_api/mod.rs

```
@@ -10,15 +10,19 @@ use crate::value::{Contract, Value};
 use alloc::collections::BTreeMap;
 use alloc::string::String;
 use alloc::vec::Vec;
+use core::convert::{TryFrom, TryInto};
 
 /// Read value under the key in the global state
 pub fn read<T>(u_ptr: UPointer<T>) -> T
 where
-    T: From<Value>,
+    T: TryFrom<Value>,
```

What was wrong with `From<Value>` ?

---

Say I want to read `i32`, `Value::Int32` is the only variant which really makes sense and others should fail the conversion. Certainly we could write `From` implementations which `panic` when they get an unexpected variant, but I thought `TryFrom` was more idiomatic.

That said, I did choose to make the error not be forwarded on (this method still returns `T` not `Result<T, _>`) because that seemed to align better with the spirit of the API. If I pass in a `UPointer<T>` then I as the programmer have some assurance that there will be a `T` in the global state, so panicking in other cases seems reasonable.

---

> Certainly we could write `From` implementations which `panic` when they get an unexpected variant.

No no, don't do that!

https://doc.rust-lang.org/beta/std/convert/trait.From.html

Specifically, "**Note: this trait must not fail**".

---

That's why I chose to go with `TryFrom` :)